core/src/main/java/ai/timefold/solver/core/impl/heuristic/selector/value/decorator/FilteringValueRangeSelector.java

```diff
@@ -0,0 +1,379 @@
+        @Nullable
+        List<Object> valueList;
+        @Nullable
+        Set<Object> entitiesSet;
+
+        AbstractFilteringValueRangeIterator(ListVariableStateSupply<Solution_> listVariableStateSupply,
+                ReachableValueMatrix reachableValueMatrix, boolean checkSourceAndDestination, boolean useValueList) {
+            this.reachableValueMatrix = Objects.requireNonNull(reachableValueMatrix);
+            this.listVariableStateSupply = listVariableStateSupply;
+            this.checkSourceAndDestination = checkSourceAndDestination;
+            this.useValueList = useValueList;
+        }
+
+        void loadValues(Object upcomingValue) {
+            this.currentUpcomingValue = upcomingValue;
+            this.entitiesSet = reachableValueMatrix.extractReachableEntities(currentUpcomingValue);
+            this.valueList = null;
+            this.valuesSet = null;
+            if (useValueList) {
+                // Load the random access list
+                valueList = Objects.requireNonNull(reachableValueMatrix.extractReachableValuesAsList(currentUpcomingValue));
+                if (valueList.isEmpty()) {
+                    noData();
+                    return;
+                }
+            } else {
+                // Load the fast access set
+                this.valuesSet = reachableValueMatrix.extractReachableValues(currentUpcomingValue);
+                if (valuesSet == null || valuesSet.isEmpty()) {
+                    noData();
+                    return;
+                }
+            }
+            currentUpcomingEntity = null;
+            if (checkSourceAndDestination) {
+                // Load the current assigned entity of the selected value
+                var position = listVariableStateSupply.getElementPosition(currentUpcomingValue);
+                if (position instanceof PositionInList positionInList) {
+                    currentUpcomingEntity = positionInList.entity();
+                }
+            }
+            this.hasData = true;
+            this.initialized = true;
+        }
+
+        void noData() {
+            this.entitiesSet = null;
+            this.valuesSet = null;
+            this.valueList = null;
+            this.currentUpcomingEntity = null;
+            this.hasData = false;
+            this.initialized = true;
+        }
+
+        boolean isEntityReachable(Object destinationValue) {
+            var destinationValid = false;
+            var sourceValid = true;
+            // Test if the destination entity accepts the selected value
+            var assignedDestinationPosition = listVariableStateSupply.getElementPosition(destinationValue);
+            if (assignedDestinationPosition instanceof PositionInList elementPosition) {
+                destinationValid = entitiesSet.contains(elementPosition.entity());
+            } else {
+                // Unassigned element is valid
+                destinationValid = true;
+            }
```

Comment:
Arguably you can save yourself some work if you return when detecting invalid destination.